### PR TITLE
feat: KPI-199 Calculator refactor to new metrics

### PR DESCRIPTION
**Pull request by FreddyJR1995**
## Jira Link

[Go to Jira](https://kpinetwork.atlassian.net/browse/KPI-199)

## Types of changes

- [ ] Hotfix (Bug detected)
- [x ] Feature (Issue from an Active Sprint)

## Checklist:

- [ x] Unit Test.
- [x ] Coverage.

## Solution
New metric calculations and answers have been added to the calculator. Also, the calculations test has been added. Other services have been modified to validate the tests.

### CODE REVIEW

File: src/service/comparison_vs_peers/comparison_vs_peers_service.py [hunk 1]
```diff
@@ -11,6 +11,10 @@ def remove_base_metrics(self, company: dict) -> None:
             "prior_actuals_revenue",
             "budget_revenue",
             "budget_ebitda",
+            "actuals_cost_of_goods",
+            "actuals_sales_of_marketing",
+            "actuals_general_and_admin",
+            "actuals_research_and_development",
```
Comment (kellycastrof) [inline src/service/comparison_vs_peers/comparison_vs_peers_service.py:17]: What about budget values?
We should only have the metrics for the reports, if we don't delete all unnecessary metrics, the download button won't work

File: src/service/calculator/calculator_repository.py [hunk 2]
```diff
@@ -55,8 +55,38 @@ def get_actuals_budget_options(self, year: int) -> list:
         return [
             self.get_metric_option("Actuals", "Revenue", "actuals_revenue", year),
             self.get_metric_option("Actuals", "Ebitda", "actuals_ebitda", year),
+            self.get_metric_option(
+                "Actuals", "Cost_of_Goods", "actuals_cost_of_goods", year
```
[the remark this answers is not in the document]
Comment (FreddyJR1995) [reply src/service/calculator/calculator_repository.py]: It was changed.

File: src/service/calculator/calculator_repository.py [hunk 3]
```diff
@@ -55,8 +55,38 @@ def get_actuals_budget_options(self, year: int) -> list:
...
+            ),
+            self.get_metric_option(
+                "Actuals", "Sales_Marketing_Cost", "actuals_sales_marketing_cost", year
```
[the remark this answers is not in the document]
Comment (FreddyJR1995) [reply src/service/calculator/calculator_repository.py]: same here.

File: src/service/calculator/calculator_repository.py [hunk 4]
```diff
@@ -55,8 +55,38 @@ def get_actuals_budget_options(self, year: int) -> list:
...
+            ),
+            self.get_metric_option(
+                "Actuals", "General_Admin_Cost", "actuals_general_admin_cost", year
+            ),
+            self.get_metric_option(
+                "Actuals",
+                "Research_Development_Cost",
```
[the remark this answers is not in the document]
Comment (FreddyJR1995) [reply src/service/calculator/calculator_repository.py]: same here.

File: src/service/calculator/calculator_repository.py [hunk 5]
```diff
@@ -84,6 +138,30 @@ def get_budget_options(self, year: int, need_prior: bool, need_next: bool) -> li
...
+                        "Budget",
+                        "General_Admin_Cost",
+                        "next_budget_general_admin_cost",
+                        next_year,
+                    ),
+                    self.get_metric_option(
+                        "Budget",
+                        "Research_Development_Cost",
+                        "next_budget_research_development_cost",
+                        next_year,
```
[the remark this answers is not in the document]
Comment (FreddyJR1995) [reply src/service/calculator/calculator_repository.py]: same here.

File: src/service/comparison_vs_peers/comparison_vs_peers_service.py (hunk 1 above)
Comment (FreddyJR1995) [reply src/service/comparison_vs_peers/comparison_vs_peers_service.py:20]: budget values were added.

File: src/service/calculator/calculator_repository.py [hunk 6]
```diff
@@ -51,12 +51,49 @@ def get_metric_option(
...
+            ),
+            self.get_metric_option(
+                "Actuals",
+                "Research & development",
+                "actuals_research_development_cost",
+                year,
+            ),
+        ]
+
+    def get_actual_budget_options(self, year: int) -> list:
```
Comment (kellycastrof) [inline src/service/calculator/calculator_repository.py:78]: Can you change the name? This functions seems to be only for budget

File: src/service/calculator/calculator_repository.py [hunk 7]
```diff
@@ -51,12 +51,49 @@ def get_metric_option(
             "alias": f"{alias}",
         }
 
-    def get_actuals_budget_options(self, year: int) -> list:
+    def get_actuals_options(self, year: int) -> list:
         return [
             self.get_metric_option("Actuals", "Revenue", "actuals_revenue", year),
             self.get_metric_option("Actuals", "Ebitda", "actuals_ebitda", year),
+            self.get_metric_option(
+                "Actuals", "Cost of good", "actuals_cost_of_goods", year
```
Comment (kellycastrof) [inline src/service/calculator/calculator_repository.py:59]: The metric is "Cost of goods"

**Review by kellycastrof — CHANGES_REQUESTED** [review]

File: src/service/calculator/calculator_repository.py [hunk 8]
```diff
@@ -71,6 +108,30 @@ def get_budget_options(self, year: int, need_prior: bool, need_next: bool) -> li
...
+                        "General & administration",
+                        "prior_budget_general_admin_cost",
+                        prior_year,
+                    ),
+                    self.get_metric_option(
+                        "Budget",
+                        "Research & development",
+                        "prior_budget_research_development_cost",
+                        prior_year,
+                    ),
```
Comment (kellycastrof) [inline src/service/calculator/calculator_repository.py:134]: We don't need these options.
Please rename the function to get_forward_budget_options

File: src/service/calculator/calculator_repository.py [hunk 9]
```diff
@@ -84,6 +145,30 @@ def get_budget_options(self, year: int, need_prior: bool, need_next: bool) -> li
...
+                        "General & administration",
+                        "next_budget_general_admin_cost",
+                        next_year,
+                    ),
+                    self.get_metric_option(
+                        "Budget",
+                        "Research & development",
+                        "next_budget_research_development_cost",
+                        next_year,
+                    ),
```
Comment (kellycastrof) [inline src/service/calculator/calculator_repository.py:171]: Delete these options

File: src/service/calculator/calculator_repository.py (hunk 6 above)
Comment (FreddyJR1995) [reply src/service/calculator/calculator_repository.py]: this was called **get_budget_options**

File: src/service/calculator/calculator_repository.py (hunk 7 above)
Comment (FreddyJR1995) [reply src/service/calculator/calculator_repository.py]: It was changed.

File: src/service/calculator/calculator_repository.py (hunk 8 above)
Comment (FreddyJR1995) [reply src/service/calculator/calculator_repository.py]: these were deleted and the funcion was renamed.

File: src/service/calculator/calculator_repository.py (hunk 9 above)
Comment (FreddyJR1995) [reply src/service/calculator/calculator_repository.py]: same here

File: src/service/calculator/calculator_report.py [hunk 10]
```diff
@@ -72,6 +72,27 @@ def calculate_metrics(self, company: dict) -> None:
...
+        company["gross_profit"] = self.calculator.calculate_gross_profit(
+            revenue, company.get("actuals_cost_of_goods")
+        )
+        company["gross_margin"] = self.calculator.calculate_gross_margin(
+            revenue, company.get("actuals_cost_of_goods")
+        )
+        company[
+            "sales_and_marketing"
+        ] = self.calculator.calculate_sales_and_marketing_of_revenue(
+            revenue, company.get("actuals_sales_of_marketing")
```
Comment (kellycastrof) [inline src/service/calculator/calculator_report.py:84]: You have different names for the same metric in report and repository, please change that.
The name should be the same.

**Review by kellycastrof — CHANGES_REQUESTED** [review]

**Review by kellycastrof — APPROVED** [review]

**Pull request merged**